source/common/http/hash_policy.cc

```diff
@@ -39,15 +42,34 @@ class HeaderHashMethod : public HashMethodImplBase {
                                     const StreamInfo::FilterStateSharedPtr) const override {
     absl::optional<uint64_t> hash;
 
-    // TODO(mattklein123): Potentially hash on all headers.
     const auto header = headers.get(header_name_);
     if (!header.empty()) {
+      absl::InlinedVector<std::string, 1> rewritten_header_values;
+      absl::InlinedVector<absl::string_view, 1> header_values;
+
+      size_t num_headers_to_hash = 1;
+      if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.hash_multiple_header_values")) {
+        num_headers_to_hash = header.size();
+        header_values.reserve(num_headers_to_hash);
+      }
+
+      for (size_t i = 0; i < num_headers_to_hash; i++) {
+        header_values.push_back(header[i]->value().getStringView());
+      }
+
       if (regex_rewrite_ != nullptr) {
-        hash = HashUtil::xxHash64(regex_rewrite_->replaceAll(header[0]->value().getStringView(),
-                                                             regex_rewrite_substitution_));
-      } else {
-        hash = HashUtil::xxHash64(header[0]->value().getStringView());
+        rewritten_header_values.reserve(num_headers_to_hash);
+        for (auto& value : header_values) {
+          rewritten_header_values.push_back(
```

Comment:
Why do we even need this vector if we only use it on L65? Can this just be a temporary?

Reply:
It serves as backing-store for the regex-written header values, so it needs to live through the sort and the vector-hash.

Reply:
Oh yeah I see, header_values is a vector of string_views. Makes sense!